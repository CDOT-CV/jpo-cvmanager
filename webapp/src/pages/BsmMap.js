import React, { useState, useEffect } from 'react'
import ReactMapGL, { Source, Layer } from 'react-map-gl'
import mbStyle from '../styles/mb_style.json'
import EnvironmentVars from '../EnvironmentVars'
import dayjs from 'dayjs'
import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider'
import { AdapterDayjs } from '@mui/x-date-pickers/AdapterDayjs'
import { DateTimePicker } from '@mui/x-date-pickers/DateTimePicker'
import TextField from '@mui/material/TextField'
import Slider from 'rc-slider'
import Select from 'react-select'
import 'rc-slider/assets/index.css'
import './css/BsmMap.css'
import {
  selectAddPoint,
  selectBsmStart,
  selectBsmEnd,
  selectBsmDateError,
  selectBsmData,
  selectBsmCoordinates,
  selectBsmFilter,
  selectBsmFilterStep,
  selectBsmFilterOffset,

  // actions
  togglePointSelect,
  clearBsm,
  updatePoints,
  updateBsmData,
  updateBsmDate,
  setBsmFilter,
  setBsmFilterStep,
  setBsmFilterOffset,
} from '../generalSlices/rsuSlice'
import { useSelector, useDispatch } from 'react-redux'

const { DateTime } = require('luxon')

const fillLayer = {
  id: 'fill',
  type: 'fill',
  source: 'polygonSource',
  layout: {},
  paint: {
    'fill-color': '#0080ff',
    'fill-opacity': 0.2,
  },
}

const outlineLayer = {
  id: 'outline',
  type: 'line',
  source: 'polygonSource',
  layout: {},
  paint: {
    'line-color': '#000',
    'line-width': 3,
  },
}

const pointLayer = {
  id: 'pointLayer',
  type: 'circle',
  source: 'pointSource',
  paint: {
    'circle-radius': 5,
    'circle-color': 'rgb(255, 164, 0)',
  },
}

function BsmMap(props) {
  const dispatch = useDispatch()

  const bsmData = useSelector(selectBsmData)
  const bsmCoordinates = useSelector(selectBsmCoordinates)
  const addPoint = useSelector(selectAddPoint)
  const startBsmDate = useSelector(selectBsmStart)
  const endBsmDate = useSelector(selectBsmEnd)
  const bsmDateError = useSelector(selectBsmDateError)
  const filter = useSelector(selectBsmFilter)
  const filterStep = useSelector(selectBsmFilterStep)
  const filterOffset = useSelector(selectBsmFilterOffset)

  const [viewport, setViewport] = useState({
    latitude: 39.7392,
    longitude: -104.9903,
    width: '100%',
    height: props.auth ? 'calc(100vh - 136px)' : 'calc(100vh - 100px)',
    zoom: 10,
  })
  const [polygonSource, setPolygonSource] = useState({
    type: 'Feature',
    geometry: {
      type: 'Polygon',
      coordinates: [],
    },
  })
  const [pointSource, setPointSource] = useState({
    type: 'FeatureCollection',
    features: [],
  })

  const [baseDate, setBaseDate] = useState(new Date(startBsmDate))
  const [startDate, setStartDate] = useState(new Date(baseDate.getTime() + 60000 * filterOffset * filterStep))
  const [endDate, setEndDate] = useState(new Date(startDate.getTime() + 60000 * filterStep))

  useEffect(() => {
    const localBaseDate = new Date(startBsmDate)
    const localStartDate = new Date(localBaseDate.getTime() + 60000 * filterOffset * filterStep)
    const localEndDate = new Date(new Date(localStartDate).getTime() + 60000 * filterStep)
    setBaseDate(localBaseDate)
    setStartDate(localStartDate)
    setEndDate(localEndDate)
  }, [startBsmDate, filterOffset, filterStep])

  const stepOptions = [
    { value: 1, label: '1 minute' },
    { value: 5, label: '5 minutes' },
    { value: 15, label: '15 minutes' },
    { value: 30, label: '30 minutes' },
    { value: 60, label: '60 minutes' },
  ]

  useEffect(() => {
    if (!startBsmDate) {
      dateChanged(new Date(), 'start')
    }
    if (!endBsmDate) {
      dateChanged(new Date(), 'end')
    }
  }, [])

  useEffect(() => {
    setPolygonSource((prevPolygonSource) => {
      return {
        ...prevPolygonSource,
        geometry: {
          ...prevPolygonSource.geometry,
          coordinates: [[...bsmCoordinates]],
        },
      }
    })

    const pointSourceFeatures = []
    if ((bsmData?.length ?? 0) > 0) {
      for (const [, val] of Object.entries([...bsmData])) {
        const bsmDate = new Date(val['properties']['time'])
        if (bsmDate >= startDate && bsmDate <= endDate) {
          pointSourceFeatures.push(val)
        }
      }
    } else {
      bsmCoordinates.forEach((point) => {
        pointSourceFeatures.push({
          type: 'Feature',
          geometry: {
            type: 'Point',
            coordinates: [...point],
          },
        })
      })
    }

    setPointSource((prevPointSource) => {
      return { ...prevPointSource, features: pointSourceFeatures }
    })
  }, [bsmCoordinates, bsmData, startDate, endDate])

  function dateChanged(e, type) {
    try {
      let mst = DateTime.fromISO(e.toISOString())
      mst.setZone('America/Denver')
      dispatch(updateBsmDate({ type, date: mst.toString() }))
    } catch (err) {
      console.error('Encountered issue updating date: ', err.message)
    }
  }

  const addPointToCoordinates = (point) => {
    if (bsmCoordinates.length > 1) {
      if (bsmCoordinates[0] === bsmCoordinates.slice(-1)[0]) {
        let tmp = [...bsmCoordinates]
        tmp.pop()
        dispatch(updatePoints([...tmp, point, bsmCoordinates[0]]))
      } else {
        dispatch(updatePoints([...bsmCoordinates, point, bsmCoordinates[0]]))
      }
    } else {
      dispatch(updatePoints([...bsmCoordinates, point]))
    }
  }

  function defaultSlider(val) {
    for (var i = 0; i < stepOptions.length; i++) {
      if (stepOptions[i].value === val) {
        return stepOptions[i].label
      }
    }
  }

  return (
    <div className="container">
      {filter ? (
        <div className="filterControl">
          <div id="timeContainer">
            <p id="timeHeader">
              {startDate.toLocaleTimeString([], {
                hour: '2-digit',
                minute: '2-digit',
              })}{' '}
              -{' '}
              {endDate.toLocaleTimeString([], {
                hour: '2-digit',
                minute: '2-digit',
              })}
            </p>
          </div>
          <div id="sliderContainer">
            <Slider
              allowCross={false}
              included={false}
              max={(new Date(endBsmDate).getTime() - baseDate.getTime()) / (filterStep * 60000)}
              value={filterOffset}
              onChange={(e) => {
                dispatch(setBsmFilterOffset(e))
              }}
            />
          </div>
          <div id="controlContainer">
            <Select
              id="stepSelect"
              options={stepOptions}
              defaultValue={filterStep}
              placeholder={defaultSlider(filterStep)}
              onChange={(e) => dispatch(setBsmFilterStep(e.value))}
            />
            <button className="searchButton" onClick={() => dispatch(setBsmFilter(false))}>
              New Search
            </button>
          </div>
        </div>
      ) : (
        <div className="control">
          <div className="buttonContainer">
            <button
              className={addPoint ? 'selected' : 'button'}
              onClick={(e) => {
                dispatch(togglePointSelect())
              }}
            >
              Add Point
            </button>
            <button
              className="button"
              onClick={(e) => {
                dispatch(clearBsm())
              }}
            >
              Clear
            </button>
          </div>
          <div className="dateContainer">
            <LocalizationProvider dateAdapter={AdapterDayjs}>
              <DateTimePicker
                label="Select start date"
                value={dayjs(startBsmDate === '' ? new Date() : startBsmDate)}
                maxDateTime={dayjs(endBsmDate === '' ? new Date() : endBsmDate)}
                onChange={(e) => {
                  dateChanged(e.toDate(), 'start')
                }}
                renderInput={(params) => <TextField {...params} />}
              />
            </LocalizationProvider>
          </div>
          <div className="dateContainer">
            <LocalizationProvider dateAdapter={AdapterDayjs}>
              <DateTimePicker
                label="Select end date"
                value={dayjs(endBsmDate === '' ? new Date() : endBsmDate)}
                minDateTime={startBsmDate === '' ? null : dayjs(startBsmDate)}
                maxDateTime={dayjs(new Date())}
                onChange={(e) => {
                  dateChanged(e.toDate(), 'end')
                }}
                renderInput={(params) => <TextField {...params} />}
              />
            </LocalizationProvider>
          </div>
          <div className="submitContainer">
            <button
              id="submitButton"
              onClick={(e) => {
                dispatch(updateBsmData())
              }}
            >
              Submit
            </button>
          </div>
          {bsmDateError ? (
            <div id="dateMessage">Date ranges longer than 24 hours are not supported due to their large data sets</div>
          ) : null}
        </div>
      )}
      <ReactMapGL
        {...viewport}
        mapboxApiAccessToken={EnvironmentVars.MAPBOX_TOKEN}
        mapStyle={mbStyle}
        onClick={
          addPoint
            ? (e) => {
                addPointToCoordinates(e.lngLat)
              }
            : null
        }
        onViewportChange={(viewport) => {
          setViewport(viewport)
        }}
      >
        {bsmCoordinates.length > 2 ? (
          <Source type="geojson" data={polygonSource}>
            <Layer {...outlineLayer} />
            <Layer {...fillLayer} />
          </Source>
        ) : null}
        <Source type="geojson" data={pointSource}>
          <Layer {...pointLayer} />
        </Source>
      </ReactMapGL>
    </div>
  )
}

export default BsmMap
