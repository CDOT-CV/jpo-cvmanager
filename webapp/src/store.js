import { configureStore } from "@reduxjs/toolkit";
import rsuReducer from "./slices/rsuSlice";
import userReducer from "./slices/userSlice";
import wzdxReducer from "./slices/wzdxSlice";
import configReducer from "./slices/configSlice";

const store = configureStore({
  reducer: {
    rsu: rsuReducer,
    user: userReducer,
    wzdx: wzdxReducer,
    config: configReducer,
  },
});

export default store;
