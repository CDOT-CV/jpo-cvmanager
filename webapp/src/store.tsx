import { Action, ThunkAction, configureStore } from '@reduxjs/toolkit'
import rsuReducer from './generalSlices/rsuSlice'
import userReducer from './generalSlices/userSlice'
import wzdxReducer from './generalSlices/wzdxSlice'
import configReducer from './generalSlices/configSlice'
import adminAddOrganizationReducer from './features/adminAddOrganization/adminAddOrganizationSlice'
import adminAddRsuReducer from './features/adminAddRsu/adminAddRsuSlice'
import adminAddUserReducer from './features/adminAdduser/adminAddUserSlice'
import adminEditOrganizationReducer from './features/adminEditOrganization/adminEditOrganizationSlice'
import adminEditRsuReducer from './features/adminEditRsu/adminEditRsuSlice'
import adminEditUserReducer from './features/adminEditUser/adminEditUserSlice'
import adminOrganizationTabReducer from './features/adminOrganizationTab/adminOrganizationTabSlice'
import adminOrganizationTabUserReducer from './features/adminOrganizationTabUser/adminOrganizationTabUserSlice'
import adminOrganizationTabRsuReducer from './features/adminOrganizationTabRsu/adminOrganizationTabRsuSlice'
import adminRsuTabReducer from './features/adminRsuTab/adminRsuTabSlice'
import adminUserTabReducer from './features/adminUserTab/adminUserTabSlice'
import menuReducer from './features/menu/menuSlice'
import intersectionMapReducer from './components/map/map-slice'
import intersectionMapLayerStyleReducer from './components/map/map-layer-style-slice'

export const setupStore = (preloadedState: any) => {
  return configureStore({
    reducer: {
      rsu: rsuReducer,
      user: userReducer,
      wzdx: wzdxReducer,
      config: configReducer,
      adminAddOrganization: adminAddOrganizationReducer,
      adminAddRsu: adminAddRsuReducer,
      adminAddUser: adminAddUserReducer,
      adminEditOrganization: adminEditOrganizationReducer,
      adminEditRsu: adminEditRsuReducer,
      adminEditUser: adminEditUserReducer,
      adminOrganizationTab: adminOrganizationTabReducer,
      adminOrganizationTabUser: adminOrganizationTabUserReducer,
      adminOrganizationTabRsu: adminOrganizationTabRsuReducer,
      adminRsuTab: adminRsuTabReducer,
      adminUserTab: adminUserTabReducer,
      menu: menuReducer,
      intersectionMap: intersectionMapReducer,
      intersectionMapLayerStyle: intersectionMapLayerStyleReducer,
    },
    preloadedState,
  })
}

type AppStore = ReturnType<typeof setupStore>
export type AppState = ReturnType<AppStore['getState']>
type AppThunk<ReturnType = void> = ThunkAction<ReturnType, AppState, unknown, Action>

export type RootState = ReturnType<ReturnType<typeof setupStore>['getState']>
