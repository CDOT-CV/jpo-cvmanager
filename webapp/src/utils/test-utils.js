/**
 * This function searches for the every react-aria SSR ids in a given HTMLElement node and replace every attribute values with a static id
 *
 * This can be usefull when you're trying to generate a snapshot of components using react-aria under the hood
 *
 * @ex :
 * ```
 * const { container } = render(<Component />);
 *
 * replaceReactAriaIds(container);
 * ```
 *
 * @param container The HTMLElement node to search for SSR ids
 */
function replaceChaoticIds(container) {
  const props = [
    {
      selector: "class",
      updateFunc: (val) =>
        val.replace(
          /css-[0-9a-z]{6}-MuiTableCell-root-MuiTablePagination-root/g,
          "css-mocked-MuiTableCell-root-MuiTablePagination-root"
        ),
    },
  ];

  container.querySelectorAll("td").forEach((item) => {
    props.forEach((prop) => {
      if (item.getAttribute(prop.selector)) {
        item.setAttribute(prop.selector, prop.updateFunc(item.getAttribute(prop.selector)));
      }
    });
  });
  container.querySelectorAll("input[aria-invalid]").forEach((item) => {
    delete item["aria-invalid"];
  });
  return container;
}

export { replaceChaoticIds };
