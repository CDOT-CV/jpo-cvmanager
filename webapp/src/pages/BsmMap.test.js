import React from 'react'
import { render } from '@testing-library/react'
import BsmMap from './BsmMap'
import { Provider } from 'react-redux'
import { setupStore } from '../store'
import { replaceChaoticIds } from '../utils/test-utils'

it('should take a snapshot', () => {
  const initialState = {
    rsu: { value: { bsmStart: new Date(2023, 2, 1, 0), bsmEnd: new Date(2023, 2, 1, 1), bsmCoordinates: [] } },
  }
  const { container } = render(
    <Provider store={setupStore(initialState)}>
      <BsmMap />
    </Provider>
  )

  expect(replaceChaoticIds(container)).toMatchSnapshot()
})
