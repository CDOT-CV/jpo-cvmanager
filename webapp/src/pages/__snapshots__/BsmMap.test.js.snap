// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="control"
    >
      <div
        class="buttonContainer"
      >
        <button
          class="button"
        >
          Add Point
        </button>
        <button
          class="button"
        >
          Clear
        </button>
      </div>
      <div
        class="dateContainer"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="mui-1"
            id="mui-1-label"
          >
            Select start date
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
              id="mui-1"
              placeholder="mm/dd/yyyy hh:mm (a|p)m"
              type="tel"
              value="03/01/2023 12:00 AM"
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
            >
              <button
                aria-label="Choose date, selected date is Mar 1, 2023"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="CalendarIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-14lo706"
              >
                <span>
                  Select start date
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="dateContainer"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="mui-2"
            id="mui-2-label"
          >
            Select end date
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
              id="mui-2"
              placeholder="mm/dd/yyyy hh:mm (a|p)m"
              type="tel"
              value="03/01/2023 01:00 AM"
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
            >
              <button
                aria-label="Choose date, selected date is Mar 1, 2023"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="CalendarIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-14lo706"
              >
                <span>
                  Select end date
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="submitContainer"
      >
        <button
          id="submitButton"
        >
          Submit
        </button>
      </div>
    </div>
    <div
      style="position: relative; width: 100%; height: calc(100vh - 100px); cursor: grab; outline: none;"
      tabindex="0"
    >
      <div
        style="position: relative; width: 100%; height: 100%;"
      >
        <div
          class=""
          style="position: absolute; width: 100%; height: 100%; overflow: hidden; visibility: inherit;"
        />
      </div>
    </div>
  </div>
</div>
`;
